### DISCUSSION TITLE
Where are the boolean operation methods for Solid and Brep geometries?

### DISCUSSION THREAD

**Discussion by simpleSketche** [iTwin/itwinjs-core#6982] [Q&A] [answered]
Completely noobie here.. I can only find the boolean operations as well as the examples for [polyface](https://www.itwinjs.org/learning/geometry/polyfaceclip/), and I assume polyface indicates mesh face only, and which doesn't support nurb surface face.

Does itwinjs provide any supports for brep boolean operations?

**Comment by bbastings** [comment]
Can you elaborate on exactly what you are trying to achieve?

Are you looking to create a geometric element whose geometry is the result of Boolean operations between solids/surfaces?

**Reply by simpleSketche** [reply]
I'm currently trying to:
1. create two breps and display them on web through the itwinjs sandbox.
2. apply boolean operations: merge, union, difference, split on the breps.
My main goal is just to try out this library and compare the results horizontally with rhinocommon library.
I'm very accustomed to rhinocommon and its workflow, so I'm also trying to figure out the correct way to work with Breps in itwinjs.

**Comment by bbastings** [comment, marked as answer]
The iTwin platform doesn't include a comprehensive public api for working with breps. Internally, our applications use a licensed 3rd party library. iTwin platform support for breps is mostly limited to display, mass properties, and snapping.

There are minimal alpha apis that support creating elements through specifying the inputs to a boolean operation that could satisfy 2, but that's pretty much it, there is no access to topology (faces, edges, vertices). From the ts side the brep is a black box.

**Reply by simpleSketche** [reply]
Thank you, this is good to know.

**Reply by simpleSketche** [reply]
@bbastings To clarify my understanding, given that the distinction between Breps and solid primitives can vary across libraries, does this imply that boolean operations are not available for solid primitives? Is converting them to polyface geometries the only way to work with boolean operations?

**Reply by bbastings** [reply]
Solid primitives must either be promoted to a brep or approximated by a polyface to perform a boolean. Going the polyface route can be undesirable when curved surfaces are involved. If booleans between solid primitives (or breps) is the only thing you are interested in, I can point you at the alpha api I previously mentioned. I don't want to steer you down that path thinking you will have full access to topology, etc. Also, given that the result of the boolean operation between solid primitives may need to be represented by a brep, it matters how you intend to use that result, i.e. is display, creating elements, extracting geometry as a polyface sufficient?

**Reply by simpleSketche** [reply]
Thanks for the clarification! The link to the Alpha doesn't seem to direct me to the correct place. It currently points to a random library containing a repo called `cors-everywhere`. Could you please check and confirm if this is the correct link? I did some digging, is [this url](https://github.com/iTwin/itwinjs-core/blob/d1f6a1a5414cac69217d47d092a31230fcf3b353/core/backend/src/test/standalone/GeometryStream.test.ts#L2543C3-L2543C59) what you are referring to?

**Reply by simpleSketche** [reply]
Also, I want to add that your responsiveness has been exceptional, I really really appreciate it!

**Reply by bbastings** [reply]
Sorry, wasn't intended to be a link, was mentioning that it's not a public api and is currently marked as alpha. I should note, this is a backend api and can not be used with a BlankConnection. Let me know if still interested.

**Reply by simpleSketche** [reply]
Yes, I’m happy to take a look, boolean operations would be sufficient to test for now! If you ever need a test user, I’d be glad to help :)

**Reply by bbastings** [reply]
Here you go...

[IModelDb.createBRepGeometry](https://github.com/iTwin/itwinjs-core/blob/f13ca14f590f564b1ce73df7b3f36c6cbb75c53f/core/backend/src/IModelDb.ts#L1409)

The input geometry is supplied in flatbuffer format.
[Example](https://github.com/iTwin/itwinjs-core/blob/f13ca14f590f564b1ce73df7b3f36c6cbb75c53f/core/backend/src/test/standalone/GeometryStream.test.ts#L2552)

There are several other tests in that same file that use the result to insert a new geometric element.
